java-frontend/src/main/java/org/sonar/java/ExecutionTimeReport.java

```diff
@@ -0,0 +1,101 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java;
+
+import java.time.Clock;
+import java.util.Comparator;
+import java.util.LinkedList;
+import java.util.stream.Collectors;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+
+public class ExecutionTimeReport {
+  private static final Logger LOG = Loggers.get(ExecutionTimeReport.class);
+
+  private static final int MAX_REPORTED_FILES = 10;
+  private static final long MIN_REPORTED_ANALYSIS_TIME_MS = 1000;
+  private static final long MIN_TOTAL_ANALYSIS_TIME_TO_REPORT_MS = 20_000;
+
+  private static class ExecutionTime {
+    private final String file;
+    private final long analysisTime;
+
+    public ExecutionTime(String file, long analysisTime) {
+      this.file = file;
+      this.analysisTime = analysisTime;
+    }
+  }
+
+  private static final Comparator<ExecutionTime> ORDER_BY_ANALYSIS_TIME_DESCENDING_AND_FILE_ASCENDING = (a, b) -> {
+    int compare = -Long.compare(a.analysisTime, b.analysisTime);
+    if (compare == 0) {
+      compare = a.file.compareTo(b.file);
+    }
+    return compare;
+  };
+
+  private final LinkedList<ExecutionTime> recordedOrderedExecutionTime = new LinkedList<>();
+  private long minRecordedOrderedExecutionTime = MIN_REPORTED_ANALYSIS_TIME_MS;
+
+  private final Clock clock;
+  private final long analysisStartTimeMS;
+  private String currentFile;
+  private long currentFileStartTimeMS;
+
+  public ExecutionTimeReport(Clock clock) {
+    this.clock = clock;
+    analysisStartTimeMS = clock.millis();
+  }
+
+  public void start(String currentFile) {
+    this.currentFile = currentFile;
+    currentFileStartTimeMS = clock.millis();
+  }
+
+  public void end() {
+    long currentAnalysisTime = clock.millis() - currentFileStartTimeMS;
+    if (currentAnalysisTime >= minRecordedOrderedExecutionTime) {
+      recordedOrderedExecutionTime.add(new ExecutionTime(currentFile, currentAnalysisTime));
+      recordedOrderedExecutionTime.sort(ORDER_BY_ANALYSIS_TIME_DESCENDING_AND_FILE_ASCENDING);
+      if (recordedOrderedExecutionTime.size() > MAX_REPORTED_FILES) {
+        recordedOrderedExecutionTime.removeLast();
+        minRecordedOrderedExecutionTime = recordedOrderedExecutionTime.stream()
+          .mapToLong(e -> e.analysisTime)
+          .min()
+          .orElse(MIN_REPORTED_ANALYSIS_TIME_MS);
+      }
+    }
+  }
+
+  public void report() {
+    long analysisEndTimeMS = clock.millis() - analysisStartTimeMS;
+    if (analysisEndTimeMS >= MIN_TOTAL_ANALYSIS_TIME_TO_REPORT_MS && !recordedOrderedExecutionTime.isEmpty()) {
+      LOG.info("Slowest analyzed files: " + toString());
+    }
+  }
+
+  @Override
+  public String toString() {
+    return recordedOrderedExecutionTime.stream()
+      .map(e -> e.file + " (" + e.analysisTime + "ms)")
+      .collect(Collectors.joining(", "));
```

Comment:
maybe on different lines?

Reply:
I know that it's not easy to read like this, but there's no other log INFO on multiple lines and it should not be confused with a stack trace. So I prefer to keep it like this.